anndata/core/anndata.py

```diff
@@ -489,38 +489,27 @@ def _init_as_actual(
         self._check_dimensions()
         self._check_uniqueness()
 
-        # raw
-        if raw is None:
-            self._raw = None
-        elif isinstance(raw, Raw):
-            self._raw = raw
-        elif isinstance(raw, cabc.Mapping):
-            # is dictionary from reading the file,
-            # nothing that is meant for a user
-            if self.isbacked:
-                raw_key = "raw/X" if "raw/X" in self.file else "raw.X"
-                if raw_key not in self.file:
-                    raise KeyError(
-                        f"Tried to check shape of raw from {self.filename}, "
-                        f"but there was no entry for 'raw/X'."
-                    )
-                if isinstance(self.file[raw_key], h5py.Group):
-                    shape = self.file[raw_key].attrs["shape"]
-                else:
-                    shape = self.file[raw_key].shape
-            else:
-                shape = raw["X"].shape
-
-            self._raw = Raw(
-                self,
-                X=raw.get("X", None),
-                var=_gen_dataframe(
-                    raw['var'], shape[1], ['var_names', 'col_names']
-                ),
-                varm=raw['varm'] if 'varm' in raw else None,
-            )
-        else:
-            raise ValueError(f'raw has unknown type {type(raw)}')
+        if self.filename:
+            assert not isinstance(
+                raw, Raw
+            ), "got raw from other adata but also filename?"
+            from ..readwrite.h5ad import _read_raw, _read_legacy_raw_h5ad
+
+            # Read raw dict from file
+            raw, raw_override = {}, raw
```

Comment:
Hmm, I did my best to do something useful here. If we should throw more errors instead and do less guessing / creative interaction, sure!

Reply:
What did old versions do in this case?

Reply:
Accept `None`, `Raw` instances, and dicts.

In case the anndata object was backed, there was some complex code for comparing `raw/X`’s shape with the shape of the `X` in the dict. Then the whole thing would be replaced …

My code now simply retrieves the backed Raw as dict and potentially overrides entries in that dict with user-specified dict members.

Reply:
> In case the anndata object was backed, there was some complex code for comparing raw/X’s shape with the shape of the X in the dict. Then the whole thing would be replaced …

Yeah, that was a bit of a nightmare.

> My code now simply retrieves the backed Raw as dict and potentially overrides entries in that dict with user-specified dict members.

There aren't a ton of tests around `.raw`. Could you make sure that cases this effects are covered, and if not, that they behave the same as they currently do?

It'd also be good to run some benchmarks (https://github.com/ivirshup/anndata-benchmarks) against this and see if it causes any regressions. There was a lot of unintended loading and copying during initialization before – especially for backed mode – which this suite helped catch.

Reply:
Your benchmark suite should be integrated into the main tests! I don’t really have time to set everything up though at the moment. Any way this could possibly make performance worse?

Reply:
> Your benchmark suite should be integrated into the main tests!

Benchmarks run on travis won't have meaningful results, we've got to have somewhere to run them before they can be part of CI. Plus they need a little cleaning up.

> Any way this could possibly make performance worse?

There have been a lot of changes around instantiation, which is where we had lots of unnecessary copies being made and stored before, so yes.

A lot of the performance implications can also be non-obvious from reading the code.